consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+        while !self.is_block_in_main_chain(&block_index) {
+            result.insert(0, block_index.clone());
+            block_index = self.get_previous_block_index(block_index.get_block_id())?;
+        }
+        Ok(result)
+    }
+
+    fn reorganize(
+        &mut self,
+        best_block_id: Option<Id<Block>>,
+        new_block_index: &BlockIndex,
+    ) -> Result<(), BlockError> {
+        let new_chain = self.get_new_chain(new_block_index)?;
+
+        // Disconnect the current chain if it is not a genesis
+        if let Some(ref best_block_id) = best_block_id {
+            let common_ancestor = self.get_previous_block_index(new_chain[0].get_block_id())?;
+            let mut current_ancestor = self
+                .db_tx
+                .get_block_index(best_block_id)?
+                .expect("Can't get block index. Inconsistent DB");
+
+            // Disconnect blocks
+            while self.is_block_in_main_chain(&current_ancestor)
+                && current_ancestor.get_block_id() != common_ancestor.get_block_id()
+            {
+                self.disconnect_tip(&mut current_ancestor)?;
+                current_ancestor =
+                    self.get_previous_block_index(current_ancestor.get_block_id())?;
+            }
+        }
+
+        // Connect the new chain
+        for block_index in new_chain {
+            self.connect_tip(&block_index, &best_block_id)?;
+        }
+        Ok(())
+    }
+
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
+        for (tx_id, tx_index) in cached_inputs.iter() {
+            self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
+        }
+        Ok(())
+    }
+
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
+
+        let tx_position =
+            TxMainChainPosition::new(&block.get_id().get(), offset_tx as u32, enc_tx.len() as u32);
+
+        assert_eq!(
+            &self.db_tx.get_mainchain_tx_by_position(&tx_position).ok().flatten().expect("Database corrupted! "),
+            tx
+        );
+
+        TxMainChainIndex::new(
+            SpendablePosition::from(tx_position),
+            tx.get_outputs().len() as u32,
+        )
+        .map_err(BlockError::from)
+    }
+
+    fn connect_transactions(&mut self, block: &Block) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in block.transactions().iter() {
+            for input in tx.get_inputs() {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
```

Comment:
Shouldn't `tx_index` here be the Index of the transaction that created the outpoint being spent? I.e. a parent of the tx we are now connecting. In line 268 we are essentially saying that the transaction that the tx being connected spends an output created by the transaction referenced by `tx_index`. In that case, `tx_index` should refer to the parent.

Reply:
You're right. And honestly I'm very upset because such trivial mistakes are not being caught with the tests. I still haven't taken a look at the tests, but I think they need to be redusted from scratch if this kind of mistake is not being caught.

Reply:
Let me double-check it